generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java

```diff
@@ -896,6 +906,35 @@ private ClassInfo classNameForUnnamedTraverse(ClassTemplateSpec enclosingClass,
     }
   }
 
+  /**
+   * Java doesn't allow inner-classnames to be same as the enclosing or ancestor class. This method takes a candidate
+   * class name for an inner class and its enclosing class and resolves to a name that doesn't conflict. It does this
+   * by adding a special separator {@link #CLASS_NAME_SUFFIX_SEPARATOR} and the provided suffix to the classname
+   * whenever a conflict is detected.
+   * The special separator is needed to ensure the new name does not conflict with classes generated from other fields.
+   * @param enclosingClass Class enclosing the inner class.
+   * @param className Candidate name for the innerclass
+   * @param suffix Suffix to add to the className if a conflict is found.
+   * @return a class name that doesn't conflict with any of the ancestor classes.
+   */
+  private String resolveInnerClassName(ClassTemplateSpec enclosingClass, String className, String suffix) {
```

Comment:
where do we guard sibling name conflict?

Reply:
There wouldn't be a conflict because we use the '$' char.

Reply:
make sense.

Reply:
Wait a minute, it just occurred to me such a counter example, let's say that domain owners copy and paste, and accidentally use "unionNameConflictArray" as a union field name. Then for the first array field "unionNameConflict", it will use "UnionNameConflictArray" since it is not conflicting with enclosing Class. Now it goes to that union field "unionNameConflictArray", we check that it is not conflicting with enclosing class either, so we just use "UnionNameConflictArray" as inner class name. Now we got sibling conflict, right?

```
record UnionNameConflictRecord {
  // Union's inner class should be UnionNameConflict, the array class will be UnionNameConflictArray$Array to avoid
  // naming conflict with record class.
  unionNameConflict: array[union[int, string]]
  // Union's inner class should be UnionNameConflictArray$Union to avoid naming conflict, and the array class will be
  // UnionNameConflictArray$UnionArray
  unionNameConflictArray: union[int, string]
} 
```

Reply:
I didn't handle this case on purpose as we cannot do it without breaking backwards compatibility. We already add Array and Map suffix for all array/map fields. And this can conflict with other fields. Any solution to fix that would change the generated class names.